### ISSUE TITLE
twisted.internet.defer.FirstError masks errors when logged

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3298]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#3298|
|Type|defect|
|Created|2008-06-18 15:44:37Z|
|Branch|https://github.com/twisted/twisted/tree/firsterror-display-3298|

An exception instance often includes extra information beyond what class it had; when errors occur this information ought to be logged. Unfortunately, `FirstError` masks this information, so Deferred Failures coming out of gatherResults are much less informative:

```
>>> from twisted.internet import defer
>>> from twisted.python import log
>>> d = defer.fail(ValueError("This failed for an important reason."))
>>> d2 = defer.gatherResults([d])
>>> d2.addErrback(log.err)
Traceback (most recent call last):
Failure: twisted.internet.defer.FirstError: FirstError(<twisted.python.failure.Failure exceptions.ValueError>, 0)
<DeferredList at 0x2aaaaf2c1050  current result: None>
>>> 
```

Notice how the log does not include the additional information ('important reason') the `ValueError` instance provides us if it would be logged directly:

```
>>> d3 = defer.fail(ValueError("This failed for an important reason."))
>>> d3.addErrback(log.err)
Traceback (most recent call last):
Failure: exceptions.ValueError: This failed for an important reason.
<Deferred at 0x2aaaaf2c1710  current result: None>
```

<details><summary>Searchable metadata</summary>

```
trac-id__3298 3298
type__defect defect
reporter__itamar itamar
priority__high high
milestone__ 
branch__branches_firsterror_display_3298 branches/firsterror-display-3298
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1213803877000000 1213803877000000
changetime__1225385884000000 1225385884000000
version__None None
owner__ 
cc__exarkun cc__itamar cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I can think of two solutions.

  1. Give `FirstError` a `__str__`/`__repr__` which includes full traceback and error information from the wrapped failure.  This way when you use `Failure.printTraceback` or print the exception, you'll see what it is wrapping.
  1. Teach the log observer about `FirstError`.  Have it reach inside and grab the wrapped information and include it in whatever is logged.  This will resemble the previous approach for logging, but when you print a `FirstError`, you'll see the same thing as you see now.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

I like [#11338](https://github.com/twisted/twisted/issues/11338) better; it's not clear to me why the logging system has to special-case behavior for specific exception types. Keep knowledge local to the places that care about it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Adding comment just to test trac emails...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

One reason I dislike the option 1 is that it results in a pretty gigantic string representation for `FirstError`s.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

If you are logging an error that you expect, for example a connection refused for a web client, then you'll either log a short version such as the string representation of the underlying exception without a traceback, or more likely a custom message: "connection was refused". in this case the verbose representation is not relevant, since you aren't logging the full Failure.

The cases in which you do log the full failure are when you have a catch all for unexpected exceptions. In these cases you want as much information as possible, because you've encountered a environmental problem or a code bug, and the more information the better. Since these cases are also relatively rare, the extra information in the log is not a problem.

As a result, I don't think the extra logging is a problem; certainly the reason I filed a bug was because of a production system having an unexpected error, and the error masking by FirstError was hiding useful information about the underlying problem (the reason string Oracle sends back on errors.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25066]) Branching to 'firsterror-display-3298'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

1.__cmp__ is missing a @since tag
 1. the second ```FirstError``` should be marked with L{} as well

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25101]) add docstrings, improve markup

refs [#3298](https://github.com/twisted/twisted/issues/3298)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_12"></a>|@radix set owner to @exarkun|
|-|-|

1. The formatting of __str__ and __repr__ is 1: weird and 2: inconsistent.

   Why brackets in one and parenthesis in the other? why is the number second in one and first in the other?

That's the only problem I see. Thanks for this branch, it helps stuff.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @radix|
|-|-|

Okay.  I changed the repr to look more like the str.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix set owner to @exarkun|
|-|-|

Ok, that looks better. Thanks. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [25200]) Merge firsterror-display-3298

Author: exarkun, itamar
Reviewer: thijs, radix
Fixes: [#3298](https://github.com/twisted/twisted/issues/3298)

Change the `__str__` and `__repr__` of `FirstError` so that they include better
and more complete information about the underlying failure (and therefore the
real cause of the exception).  This means that when these errors are logged, the
log will now include information about the exception which caused the `FirstError`.

The deprecated tuple-like interface is also removed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
